File: ios/myDriverDriver7/BackgroundLocationModule.swift
import Foundation
import CoreLocation
import React

@objc(BackgroundLocationModule)
class BackgroundLocationModule: RCTEventEmitter, CLLocationManagerDelegate {
    
    private var locationManager: CLLocationManager?
    private var isTracking = false
    private var hasListeners = false
    
    override init() {
        super.init()
        setupLocationManager()
    }
    
    override func supportedEvents() -> [String]! {
        return ["BackgroundLocationUpdate"]
    }
    
    override func startObserving() {
        hasListeners = true
    }
    
    override func stopObserving() {
        hasListeners = false
    }
    
    @objc
    override static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    private func setupLocationManager() {
        locationManager = CLLocationManager()
        locationManager?.delegate = self
        locationManager?.desiredAccuracy = kCLLocationAccuracyBest
        locationManager?.distanceFilter = 0 // Update even if driver hasn't moved
        
        // Request permissions
        locationManager?.requestWhenInUseAuthorization()
        locationManager?.requestAlwaysAuthorization()
    }
    
    @objc
    func startBackgroundLocation(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        
        print("🚀 Starting iOS background location tracking...")
        
        guard let locationManager = locationManager else {
            reject("LOCATION_MANAGER_ERROR", "Location manager not initialized", nil)
            return
        }
        
        // Check authorization status
        let authStatus = locationManager.authorizationStatus
        guard authStatus == .authorizedAlways || authStatus == .authorizedWhenInUse else {
            reject("PERMISSION_DENIED", "Location permission not granted", nil)
            return
        }
        
        if isTracking {
            print("📍 Background location tracking already running")
            resolve("Already tracking")
            return
        }
        
        // Configure for background location
        locationManager.allowsBackgroundLocationUpdates = true
        locationManager.pausesLocationUpdatesAutomatically = false
        
        // Start location updates
        locationManager.startUpdatingLocation()
        
        // Start significant location changes for background
        locationManager.startMonitoringSignificantLocationChanges()
        
        isTracking = true
        print("✅ iOS background location tracking started successfully")
        resolve("Background location tracking started")
    }
    
    @objc
    func stopBackgroundLocation() {
        print("🛑 Stopping iOS background location tracking...")
        
        guard let locationManager = locationManager else {
            return
        }
        
        locationManager.stopUpdatingLocation()
        locationManager.stopMonitoringSignificantLocationChanges()
        locationManager.allowsBackgroundLocationUpdates = false
        
        isTracking = false
        print("✅ iOS background location tracking stopped successfully")
    }
    
    @objc
    func isBackgroundLocationRunning(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        resolve(isTracking)
    }
    
    // MARK: - CLLocationManagerDelegate
    
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let location = locations.last, hasListeners else { return }
        
        print("📍 iOS location update: \(location.coordinate.latitude), \(location.coordinate.longitude)")
        
        let locationData: [String: Any] = [
            "latitude": location.coordinate.latitude,
            "longitude": location.coordinate.longitude,
            "accuracy": location.horizontalAccuracy,
            "altitude": location.altitude,
            "speed": location.speed >= 0 ? location.speed : 0,
            "heading": location.course >= 0 ? location.course : 0,
            "timestamp": location.timestamp.timeIntervalSince1970 * 1000
        ]
        
        sendEvent(withName: "BackgroundLocationUpdate", body: locationData)
        print("📤 iOS location update sent to React Native")
    }
    
    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        print("❌ iOS location manager error: \(error.localizedDescription)")
    }
    
    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        print("📍 iOS location authorization status changed: \(status.rawValue)")
        
        switch status {
        case .authorizedAlways, .authorizedWhenInUse:
            print("✅ iOS location permission granted")
        case .denied, .restricted:
            print("❌ iOS location permission denied")
            if isTracking {
                stopBackgroundLocation()
            }
        case .notDetermined:
            print("⏳ iOS location permission not determined")
            locationManager?.requestAlwaysAuthorization()
        @unknown default:
            print("❓ iOS unknown location authorization status")
        }
    }
}

File: android/app/src/main/java/com/mydriverdriver7/BackgroundLocationModule.kt
package com.mydriverdriver7

import android.Manifest
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Build
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import androidx.core.app.ActivityCompat
import androidx.core.app.NotificationCompat
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule

class BackgroundLocationModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    companion object {
        private const val TAG = "BackgroundLocationModule"
        private const val CHANNEL_ID = "background_location_channel"
        private const val NOTIFICATION_ID = 1
        private const val LOCATION_UPDATE_INTERVAL = 5000L // 5 seconds
        private const val LOCATION_UPDATE_DISTANCE = 0f // Update even if driver hasn't moved
    }

    private var locationManager: LocationManager? = null
    private var locationListener: LocationListener? = null
    private var isTracking = false

    override fun getName(): String {
        return "BackgroundLocationModule"
    }

    init {
        createNotificationChannel()
    }

    @ReactMethod
    fun startBackgroundLocation(promise: Promise) {
        try {
            Log.d(TAG, "Starting background location tracking...")
            
            if (isTracking) {
                Log.d(TAG, "Background location tracking already running")
                promise.resolve("Already tracking")
                return
            }

            val context = reactApplicationContext
            locationManager = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager

            // Check permissions
            if (ActivityCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED &&
                ActivityCompat.checkSelfPermission(context, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                promise.reject("PERMISSION_DENIED", "Location permission not granted")
                return
            }

            // Create location listener
            locationListener = object : LocationListener {
                override fun onLocationChanged(location: Location) {
                    Log.d(TAG, "📍 Location update: ${location.latitude}, ${location.longitude}")
                    sendLocationUpdate(location)
                }

                override fun onStatusChanged(provider: String?, status: Int, extras: Bundle?) {}
                override fun onProviderEnabled(provider: String) {}
                override fun onProviderDisabled(provider: String) {}
            }

            // Start location updates
            locationManager?.requestLocationUpdates(
                LocationManager.GPS_PROVIDER,
                LOCATION_UPDATE_INTERVAL,
                LOCATION_UPDATE_DISTANCE,
                locationListener!!
            )

            // Also request from network provider as backup
            if (locationManager?.isProviderEnabled(LocationManager.NETWORK_PROVIDER) == true) {
                locationManager?.requestLocationUpdates(
                    LocationManager.NETWORK_PROVIDER,
                    LOCATION_UPDATE_INTERVAL,
                    LOCATION_UPDATE_DISTANCE,
                    locationListener!!
                )
            }

            // Start foreground service
            startForegroundService()
            
            isTracking = true
            Log.d(TAG, "✅ Background location tracking started successfully")
            promise.resolve("Background location tracking started")

        } catch (e: Exception) {
            Log.e(TAG, "❌ Error starting background location tracking", e)
            promise.reject("START_ERROR", "Failed to start background location tracking: ${e.message}")
        }
    }

    @ReactMethod
    fun stopBackgroundLocation() {
        try {
            Log.d(TAG, "Stopping background location tracking...")
            
            locationListener?.let { listener ->
                locationManager?.removeUpdates(listener)
            }
            
            stopForegroundService()
            
            isTracking = false
            locationListener = null
            locationManager = null
            
            Log.d(TAG, "✅ Background location tracking stopped successfully")
        } catch (e: Exception) {
            Log.e(TAG, "❌ Error stopping background location tracking", e)
        }
    }

    @ReactMethod
    fun isBackgroundLocationRunning(promise: Promise) {
        promise.resolve(isTracking)
    }

    private fun sendLocationUpdate(location: Location) {
        try {
            val params = Arguments.createMap().apply {
                putDouble("latitude", location.latitude)
                putDouble("longitude", location.longitude)
                putDouble("accuracy", location.accuracy.toDouble())
                putDouble("altitude", location.altitude)
                putDouble("speed", location.speed.toDouble())
                putDouble("heading", location.bearing.toDouble())
                putDouble("timestamp", location.time.toDouble())
            }

            reactApplicationContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit("BackgroundLocationUpdate", params)

            Log.d(TAG, "📤 Location update sent to React Native")
        } catch (e: Exception) {
            Log.e(TAG, "❌ Error sending location update to React Native", e)
        }
    }

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                CHANNEL_ID,
                "Background Location",
                NotificationManager.IMPORTANCE_LOW
            ).apply {
                description = "Driver location tracking"
                setShowBadge(false)
            }

            val notificationManager = reactApplicationContext.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            notificationManager.createNotificationChannel(channel)
        }
    }

    private fun startForegroundService() {
        val intent = Intent(reactApplicationContext, MainActivity::class.java)
        val pendingIntent = PendingIntent.getActivity(
            reactApplicationContext,
            0,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )

        val notification = NotificationCompat.Builder(reactApplicationContext, CHANNEL_ID)
            .setContentTitle("Driver Location Tracking")
            .setContentText("Tracking your location for ride requests")
            .setSmallIcon(android.R.drawable.ic_menu_mylocation)
            .setContentIntent(pendingIntent)
            .setOngoing(true)
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .build()

        // Start foreground service
        val serviceIntent = Intent(reactApplicationContext, BackgroundLocationService::class.java)
        serviceIntent.putExtra("notification", notification)
        
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            reactApplicationContext.startForegroundService(serviceIntent)
        } else {
            reactApplicationContext.startService(serviceIntent)
        }
    }

    private fun stopForegroundService() {
        val serviceIntent = Intent(reactApplicationContext, BackgroundLocationService::class.java)
        reactApplicationContext.stopService(serviceIntent)
    }
}
